aten/src/ATen/core/thread_pool.cpp

```diff
@@ -3,23 +3,113 @@
 
 namespace c10 {
 
-ThreadPool::ThreadPool() { /* intended to leave empty */ }
+TaskThreadPoolBase::~TaskThreadPoolBase() { /* intended to leave empty */ }
 
-ThreadPool::~ThreadPool() { /* intended to leave empty */ }
+ThreadPool::ThreadPool(std::size_t pool_size, int numa_node_id)
+    : threads_(pool_size),
+      running_(true),
+      complete_(true),
+      available_(pool_size),
+      total_(pool_size),
+      numa_node_id_(numa_node_id) {
+  for (std::size_t i = 0; i < pool_size; ++i) {
+    threads_[i] = std::thread(std::bind(&ThreadPool::main_loop, this, i));
+  }
+}
+
+ThreadPool::~ThreadPool() {
+  // Set running flag to false then notify all threads.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    running_ = false;
+    condition_.notify_all();
+  }
+
+  try {
+    for (auto& t : threads_) {
+      t.join();
+    }
+  } catch (const std::exception&) {
+  }
+}
+
+size_t ThreadPool::size() const {
+  return threads_.size();
+}
+
+size_t ThreadPool::numAvailable() const {
+  return available_;
+}
 
-void ThreadPool::schedule(std::function<void(void)> task) {
-  tasks.push(task);
+void ThreadPool::run(const std::function<void()>& func) {
+  runTask(func);
 }
 
-void ThreadPool::workOnTasksUntilCompleted(
-    c10::intrusive_ptr<ivalue::Future> future) {
-  while (!future->completed()) {
-    auto task = tasks.front();
-    tasks.pop();
-    task();
+void ThreadPool::waitWorkComplete() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (!complete_) {
+    completed_.wait(lock);
   }
 }
 
-ThreadPool global_work_queue;
+void ThreadPool::waitWorkComplete(c10::intrusive_ptr<ivalue::Future> future) {
+  // TODO: wait on future
+  waitWorkComplete();
```

Comment:
A couple nits:

1. Why change the name of `workOnTasksUntilCompleted`? Even though that is not what the method does now, it reflects the design as we want it to work. We can easily forget the places where it is valid to have the thread join the worker pool if we rename it to something else and forget to name it back. It is better to leave it as is, with a note for how it should work in the future.

2. This implementation will livelock when there is a continual stream of work coming into the threadpool. If the thread is going to sleep then it should arrange for it to be woken up when Future is completed.